File: src/collectors/podman.ts
import { getPodmanSocket } from '../utils/socket';
import { calculateRate } from '../utils/calc';
import type { ContainerStat } from '../types';

export async function collectContainerMetrics(): Promise<ContainerStat[]> {
  const socketPath = await getPodmanSocket();
  if (!socketPath) return [];

  try {
    const response = await fetch('http://d/v5.0.0/libpod/containers/stats?stream=false', {
      unix: socketPath,
    });

    if (!response.ok) {
      throw new Error(`Podman API Error: ${response.status}`);
    }

    const json = await response.json() as any;
    
    
    let rawStats: any[] = [];

    if (Array.isArray(json.Stats)) {
      rawStats = json.Stats;
    } 
    else if (json.Stats && typeof json.Stats === 'object') {
      rawStats = Object.values(json.Stats);
    }
    else if (Array.isArray(json)) {
      rawStats = json;
    }

    if (rawStats.length === 0 && json.Stats) {
      console.log('⚠️ [调试] Stats 结构异常:', JSON.stringify(json).substring(0, 100));
    }

    const timestamp = Date.now();
    const results: ContainerStat[] = [];

    for (const s of rawStats) {
      const idRaw = s.ID || s.id || s.ContainerID || '';
      const nameRaw = s.Name || s.name || 'unknown';
      
      if (!idRaw) continue;

      const id = idRaw.substring(0, 12);

      let rxTotal = 0;
      let txTotal = 0;

      if (s.Network) {
        for (const ifaceName in s.Network) {
          const iface = s.Network[ifaceName];
          rxTotal += iface.RxBytes || 0;
          txTotal += iface.TxBytes || 0;
        }
      } else if (s.NetInput !== undefined) {
        rxTotal = s.NetInput || 0;
        txTotal = s.NetOutput || 0;
      }

      const rxRate = calculateRate(`container_${id}_rx`, rxTotal, timestamp);
      const txRate = calculateRate(`container_${id}_tx`, txTotal, timestamp);

      let cpuVal = 0;
      if (typeof s.CPU === 'string') {
        cpuVal = parseFloat(s.CPU.replace('%', ''));
      } else {
        cpuVal = s.CPU || 0;
      }

      const memUsage = s.MemUsage || 0;
      const memLimit = s.MemLimit || 0;
      let memPerc = s.MemPerc || 0;
      if (typeof s.MemPerc === 'string') {
        memPerc = parseFloat(s.MemPerc.replace('%', ''));
      }

      // 解析容器状态
      let state: ContainerStat['state'] = 'unknown';
      const rawState = s.State || s.status || s.Status || '';
      if (typeof rawState === 'string') {
        const stateLower = rawState.toLowerCase();
        if (stateLower.includes('running')) state = 'running';
        else if (stateLower.includes('exited') || stateLower.includes('stopped')) state = 'exited';
        else if (stateLower.includes('paused')) state = 'paused';
        else if (stateLower.includes('created')) state = 'created';
      }

      results.push({
        id: id,
        name: nameRaw,
        image: s.Image || s.image || 'unknown',
        state: state,

        cpuPercent: cpuVal,
        
        memory: {
          usage: memUsage,
          limit: memLimit,
          usagePercent: memPerc,
        },

        network: {
          rxRate,
          txRate,
          rxTotal,
          txTotal
        }
      });
    }

    return results;

  } catch (error: any) {
    console.warn(`⚠️ [Podman 采集器] 失败: ${error.message}`);
    return [];
  }
}

File: src/collectors/podman.rs
use crate::collectors::host::MetricsState;
use crate::podman::{stats::NumberOrString, PodmanClient};
use crate::types::{ContainerMemoryStat, ContainerNetworkStat, ContainerStat};
use anyhow::Result;

pub async fn collect_container_metrics(
    client: &PodmanClient,
    rate: &mut MetricsState,
) -> Result<Vec<ContainerStat>> {
    if !client.socket_exists() {
        tracing::debug!(
            socket = %client.socket_path().display(),
            "podman: socket not found, skip container metrics"
        );
        return Ok(Vec::new());
    }

    let timestamp = crate::collectors::host::now_ms();
    let containers = client.container_list().await?;
    tracing::debug!(total = containers.len(), "podman: collecting container metrics");
    let mut results = Vec::new();

    for container in containers {
        if !matches!(container.state.as_deref(), Some("running") | Some("Running") | None) {
            tracing::debug!(
                container_id = %container.id,
                state = ?container.state,
                "podman: skip non-running container"
            );
            continue;
        }

        let stats = match client.container_stats_raw(&container.id).await {
            Ok(stats) => stats,
            Err(error) => {
                tracing::warn!(
                    container_id = %container.id,
                    %error,
                    "podman: stats fetch failed, aborting collect cycle"
                );
                return Err(error);
            }
        };
        for stat in stats {
            let id_raw = stat.id.unwrap_or_else(|| container.id.clone());
            let id = id_raw.chars().take(12).collect::<String>();
            let name = stat.name.unwrap_or_else(|| container_name(&container));

            let (mut rx_total, mut tx_total) = (0, 0);
            if let Some(network) = stat.network {
                for (iface_name, iface) in network {
                    if iface_name == "lo" {
                        continue;
                    }
                    rx_total += iface.rx_bytes.unwrap_or(0);
                    tx_total += iface.tx_bytes.unwrap_or(0);
                }
            } else {
                rx_total = stat.net_input.unwrap_or(0);
                tx_total = stat.net_output.unwrap_or(0);
            }

            let cpu_percent = stat.cpu.as_ref().map(NumberOrString::as_f64).unwrap_or(0.0);
            let usage_percent = stat.mem_perc.as_ref().map(NumberOrString::as_f64).unwrap_or(0.0);

            results.push(ContainerStat {
                id: id.clone(),
                name,
                cpu_percent,
                memory: ContainerMemoryStat {
                    usage: stat.mem_usage.as_ref().map(NumberOrString::as_u64).unwrap_or(0),
                    limit: stat.mem_limit.as_ref().map(NumberOrString::as_u64).unwrap_or(0),
                    usage_percent,
                },
                network: ContainerNetworkStat {
                    rx_rate: rate.calculate(&format!("container_{id}_rx"), rx_total, timestamp),
                    tx_rate: rate.calculate(&format!("container_{id}_tx"), tx_total, timestamp),
                    rx_total,
                    tx_total,
                },
            });
        }
    }

    tracing::debug!(collected = results.len(), "podman: container metrics collected");
    Ok(results)
}

fn container_name(container: &crate::podman::stats::PodmanContainerSummary) -> String {
    container
        .name
        .clone()
        .or_else(|| container.names.first().cloned())
        .map(|name| name.trim_start_matches('/').to_string())
        .unwrap_or_else(|| "unknown".to_string())
}
